src/context.rs

```diff
@@ -2600,3 +2600,173 @@ impl ContextWriter {
+}
+
+/* Symbols for coding magnitude class of nonzero components */
+const MV_CLASSES:usize = 11;
+
+// MV Class Types
+const MV_CLASS_0: usize = 0;   /* (0, 2]     integer pel */
+const MV_CLASS_1: usize = 1;   /* (2, 4]     integer pel */
+const MV_CLASS_2: usize = 2;   /* (4, 8]     integer pel */
+const MV_CLASS_3: usize = 3;   /* (8, 16]    integer pel */
+const MV_CLASS_4: usize = 4;   /* (16, 32]   integer pel */
+const MV_CLASS_5: usize = 5;   /* (32, 64]   integer pel */
+const MV_CLASS_6: usize = 6;   /* (64, 128]  integer pel */
+const MV_CLASS_7: usize = 7;   /* (128, 256] integer pel */
+const MV_CLASS_8: usize = 8;   /* (256, 512] integer pel */
+const MV_CLASS_9: usize = 9;   /* (512, 1024] integer pel */
+const MV_CLASS_10: usize = 10; /* (1024,2048] integer pel */
+
+const CLASS0_BITS: usize = 1; /* bits at integer precision for class 0 */
+const CLASS0_SIZE: usize = (1 << CLASS0_BITS);
+const MV_OFFSET_BITS: usize = (MV_CLASSES + CLASS0_BITS - 2);
+const MV_BITS_CONTEXTS: usize = 6;
+const MV_FP_SIZE: usize = 4;
+
+const MV_MAX_BITS: usize = (MV_CLASSES + CLASS0_BITS + 2);
+const MV_MAX: usize = ((1 << MV_MAX_BITS) - 1);
+const MV_VALS: usize = ((MV_MAX << 1) + 1);
+
+const MV_IN_USE_BITS: usize = 14;
+const MV_UPP: i32 = (1 << MV_IN_USE_BITS);
+const MV_LOW: i32 = (-(1 << MV_IN_USE_BITS));
+
+
+pub struct nmv_component {
+  classes_cdf: [u16; MV_CLASSES + 1],
+  class0_fp_cdf: [[u16; MV_FP_SIZE + 1]; CLASS0_SIZE],
+  fp_cdf: [u16; MV_FP_SIZE + 1],
+  sign_cdf: [u16; 2 + 1],
+  class0_hp_cdf: [u16; 2 + 1],
+  hp_cdf: [u16; 2 + 1],
+  class0_cdf: [u16; CLASS0_SIZE + 1],
+  bits_cdf: [[u16; 2 + 1]; MV_OFFSET_BITS],
+}
+
+pub struct nmv_context {
+  joints_cdf: [u16; MV_JOINTS + 1],
+  comps: [nmv_component; 2],
+}
+
+#[derive(Copy, Clone, Debug, PartialEq, PartialOrd)]
+pub enum MvSubpelPrecision {
+  MV_SUBPEL_NONE = -1,
+  MV_SUBPEL_LOW_PRECISION = 0,
+  MV_SUBPEL_HIGH_PRECISION,
+}
+
+pub struct MV {
+  row : i16,
+  col : i16,
+}
+
+#[inline(always)]
+pub fn av1_get_mv_joint(mv: &MV) -> MV_JOINT_TYPE {
+  if mv.row == 0 {
+    if mv.col == 0 { MV_JOINT_TYPE::MV_JOINT_ZERO } else { MV_JOINT_TYPE::MV_JOINT_HNZVZ }
+  } else {
+    if mv.col == 0 { MV_JOINT_TYPE::MV_JOINT_HZVNZ } else { MV_JOINT_TYPE::MV_JOINT_HNZVNZ }
+  }
+}
+#[inline(always)]
+pub fn mv_joint_vertical(joint_type: MV_JOINT_TYPE) -> bool {
+  joint_type == MV_JOINT_TYPE::MV_JOINT_HZVNZ || joint_type == MV_JOINT_TYPE::MV_JOINT_HNZVNZ
+}
+#[inline(always)]
+pub fn mv_joint_horizontal(joint_type: MV_JOINT_TYPE ) -> bool {
+  joint_type == MV_JOINT_TYPE::MV_JOINT_HNZVZ || joint_type == MV_JOINT_TYPE::MV_JOINT_HNZVNZ
+}
+#[inline(always)]
+pub fn mv_class_base(mv_class: usize) -> u32 {
+  if mv_class != MV_CLASS_0 {
+    (CLASS0_SIZE << (mv_class as usize + 2)) as u32 }
+  else { 0 }
+}
+#[inline(always)]
+// If n != 0, returns the floor of log base 2 of n. If n == 0, returns 0.
+pub fn log_in_base_2(n: u32) -> u8 {
```

Comment:
I prepared for you a [generic ilog](https://github.com/xiph/rav1e/pull/448). You may use it :)

Reply:
I will...!